### ISSUE TITLE
Error when resizing Canvas

### ISSUE THREAD

**Issue by moaiii** [molleindustria/p5.play#115]
This code will paint a red dot in the middle of a square canvas with simple P5.js - when resize the red dot stays in the middle of the screen.  If you include the p5.play.js library in the head of index.html, change nothing in 'sketch.js' the red dot will 'drift'. 

In short P5 is messing with the resizeEvent :(

**sketch.js**

```
function setup() {
  createCanvas(windowWidth, windowWidth);
}

function draw() {
    background(0);
    noStroke();
    fill(255, 0, 0)
    ellipse(windowWidth/2, windowWidth/2, 10, 10);
}

function windowResized() {
  resizeCanvas(windowWidth, windowWidth);
}
```

**index.html**

```
<html>
<head>
  <meta charset="utf-8">

  <!-- Libraries -->
  <script src="js/libraries/p5.js"></script>
  <!-- <script src="js/libraries/p5.play.js"></script>  -->

  <!-- Javascript functionality -->
  <script src="js/sketch.js"></script>
</head>

<body>
</body>
</html>
```

**Comment by aimed-yujin** [comment]
I'm having same issue, it truely p5 play problem.

**Comment by bnew** [comment]
does anyone have any ideas on what causes this problem?

**Comment by bnew** [comment]
@islemaster  thank you for maintaining this library. It's been hugely helpful for me.

I think the issue here is that the camera is always tracked as x and y positions, rather than as a percentage of browser width or browser height. So when the browser is resized, those x and y positions will not remain the same. 

My thought would be to add a method to set the camera position instead of accessing the property directly. That way the camera's position can be saved when changed within a game and restored on browser resize. 

This isn't ideal but what I'm thinking of is:

let percentX; 
let percentY;

function setCameraPosition(x,y){
cameraSave()
camera.position.x = x;
camera.position.y = y;
}
function windowResized() {
resizeCanvas(windowWidth, windowHeight);
cameraRestore();
}

function cameraSave(){
    percentX = camera.position.x/width;
    percentY = camera.position.y/height;
}
    
function cameraRestore(){
    camera.position.x = percentX * width;
    camera.position.y = percentY * height;
}


If you have any ideas on a more elegant way to patch this, please let me know. This method is definitely less convenient than accessing the camera properties directly.

**Comment by islemaster** [comment]
Hey @bnew @moaiii, sorry I've been idle on this issue.  I'll carve out some time to take a look tonight.  Thanks!

**Comment by islemaster** [comment]
After investigating this I'm convinced that the current behavior is by design.  I'm putting together an example to illustrate that and an ideal workaround for the example you provided.

That said, I understand this is surprising behavior - it's odd that vanilla p5 and p5+p5.play behave differently in this relatively simply case.  The root cause is not exactly a change to the resize event, as @moaiii suggested - it's the presence of a camera by default, which applies a matrix transform to all `draw()` operations, and a choice of default camera position that is a weird compromise.

It begs the question - should the p5.play camera become entirely opt-in?  It would be a breaking change to existing projects, but it might make more sense.

Anyway, I'll get the explanation of current behavior posted in a half-hour or so.

**Comment by islemaster** [comment]
Here's an annotated example of how to work with the camera in p5.play.

_**TL;DR** If you want all drawing to be canvas-relative like vanilla p5.js, [call `camera.off()`](http://molleindustria.github.io/p5.play/docs/classes/Camera.html#method-off) at the top of your `draw()` function._

I've created an expanded example that shows how the camera works in p5.play, and why the red dot drifts in @moaiii's example, above.  Of course the example is best explored interactively - [the complete source is here](https://gist.github.com/islemaster/02b4ad929fcd9ac44073b63bbeb7b2da).  To aid reading along, here's an initial state of the sketch:

![image](https://user-images.githubusercontent.com/1615761/33593209-394de150-d943-11e7-9095-9a99bb5d7a59.png)

Let's get started.  Here's the opening boilerplate: Set up the canvas, resize it when the window is resized.

```js
function setup() {
    createCanvas(windowWidth - 30, windowWidth - 30);
}

function windowResized() {
  resizeCanvas(windowWidth - 30, windowWidth - 30);
}

function draw() {
  background(0);
```

p5.play creates a camera object by default, which has a center and a zoom value.  This camera starts pointed at the center of the canvas, even though the world model puts `0,0` at the canvas top-left.  Therefore the camera starts with a nonzero position.

```js
  // Green ring and text for camera position
  stroke(0, 255, 0);
  strokeWeight(2);
  noFill();
  ellipse(camera.position.x, camera.position.y, 10, 10);
  noStroke();
  fill(0, 255, 0);
  text('Camera pointed at world(' + camera.position.x + ',' + camera.position.y + ')',
    camera.position.x + 8, camera.position.y - 8);
```

![out](https://user-images.githubusercontent.com/1615761/33593665-347a644e-d945-11e7-8d2d-8108d20a636e.gif)

The camera ensures the center of the canvas is always the camera position. Here are some purple rectangles with fixed world-positions for reference. Notice that their coordinates do not change, and their position is fixed relative to the camera position.
```js
  for (var i = 0 ; i < 4; i++) {
    for (var j = 0; j < 4; j++) {
      stroke(120, 0, 120);
      strokeWeight(1);
      noFill();
      rect(200*i, 200*j, 200, 200);
      noStroke();
      fill(120, 0, 120);
      text('world(' + 200*i + ',' + 200*j + ')', 200*i+4, 200*j+12);
    }
  }
```

In the example case @moaiii gave, the red dot is drawn at the center of the canvas in the initial canvas size, but it drifts as the window resizes. This is because:
- With p5.play, by default the dot's position is interpreted in world coordinates, not canvas coordinates. --
- The dot's position is being recomputed every frame, as half of the window width.
- Therefore, the dot moves around the world as the window resizes.
- Because the camera is always centered on the canvas, and the camera has a fixed position in the world, and the dot is moving around the world, the dot "drifts."

```js
  noStroke();
  fill(255, 0, 0);
  ellipse(width/2, width/2, 6, 6);
  text('Original dot at world(' + width/2 + ',' + width/2 + ')',
    width/2 + 12, width/2 + 4);
```

To draw canvas-relative with p5.play (which seems to be the default behavior for p5.js) you have to call [`camera.off()`](http://molleindustria.github.io/p5.play/docs/classes/Camera.html#method-off).

```js
  camera.off();
```

To illustrate, here's a ring at canvas center, in yellow.

```js
  stroke(255, 255, 0);
  strokeWeight(2);
  noFill();
  ellipse(width/2, width/2, 15, 15);
  noStroke();
  fill(255, 255, 0);
  text('Ring at canvas(' + width/2 + ',' + width/2 + ')',
    width/2 + 8, width/2 + 16);
```

And here's a yellow rectangle near canvas-edge

```js
  stroke(255, 255, 0);
  strokeWeight(2);
  noFill();
  rect(20, 20, width - 40, width - 40);
  noStroke();
  fill(255, 255, 0);
  text('canvas(20,20)',
    24, 32);
  textAlign('right');
  text('canvas(' + (width - 40) + ',' + (width - 40) + ')',
    width - 24, width - 24);
  textAlign('left');
```
You can call [`camera.on()`](http://molleindustria.github.io/p5.play/docs/classes/Camera.html#method-on) to go back to drawing in world-coordinates. You can do this as many times as you need to in your draw loop.  Here's a blue dot at `150,150` in world-coordinates.

```js
  camera.on();
  noStroke();
  fill(0, 0, 255);
  ellipse(150, 150, 6, 6);
  text('world(150, 150)',
    150 + 12, 150 + 4);
}
```

![out](https://user-images.githubusercontent.com/1615761/33593332-cb1a807a-d943-11e7-9abc-6736bc18d8da.gif)

Again, [here's the complete example](https://gist.github.com/islemaster/02b4ad929fcd9ac44073b63bbeb7b2da).

 I hope that's helpful.  Does the red dot drift make sense now?  You're mixing coordinate systems - using a canvas-space value (the canvas size) as input to a world-space operation (drawing the ellipse while the camera is on).

When you're starting out with p5.play it's easy to miss this world-space vs canvas-space distinction, because, if you never resize the canvas or move the camera, the coordinate spaces line up.  A camera move or a canvas resize causes those coordinate spaces to diverge, and you have to think about them.  It's actually really useful to have these two spaces for making games though - if you want to move the camera around your game world and zoom in and out, but still have UI elements like "score" that always show up in the corner of the screen, you'll want to use world-space for your game world and canvas-space for the UI elements.

**Comment by moaiii** [comment]
@islemaster excellent demonstration! Thanks for all your help

**Comment by swirly** [comment]
Thank you very much @islemaster .
Definitively, camera should really be opt-in.
I was in the process of learning p5play, and resize is something usual when you know a bit p5js (but not yet p5play).
It may trouble a lot of people. I shall certainly not be the last to stuble upon this.
If you consider p5play is not an industrial framework but something to learn, you may confuse peoples.
Camera should be opt in and this case shall be in the documentation
Hope this helps.

**Comment by UrsulaMedium** [comment]
Thanks for the solution!
Is it possible to add the Camera behaviour to the documentation? (or make the camera opt-in indeed).  
I spent many many hours figuring what I'm doing wrong until I switched off P5 Play library.
